### ISSUE TITLE
No unit tests

### ISSUE THREAD

**Issue by masasin** [benkroop/robinhood#3]
There are no unit tests.

Also, I would like to help. Are there any planned features, or things that have not been implemented yet?

**Comment by benkroop** [comment]
That's right. I actually had an offline discussion yesterday about unit tests and decided to deprioritize them. Because all computation is done by the server, the unit tests would mostly just be testing Robinhood's endpoints. I'm planning to write a test suite for this package that I can run nightly to verify that all functionality is still supported by the API, but I have other features I'd like to write first.

There are definitely other features I'd like to add. I don't have a full list written up yet, but this is a start:
- Get account balances 
- Get list of open positions
- Initiate ACH transfer to / from bank account
- Add bank account records 
- Load a list of linked bank accounts to simplify the initiate ACH transfer function
- Get data about markets, including the next opening time and closing time: 
```python
>>> r.session.get('https://api.robinhood.com/markets/')
{"previous":null,"results":[{"website":"www.otcmarkets.com","city":"New York","name":"Otc Markets","url":"https://api.robinhood.com/markets/OTCM/","country":"United States of America","todays_hours":"https://api.robinhood.com/markets/OTCM/hours/2015-09-08/","operating_mic":"OTCM","acronym":"OTCM","timezone":"US/Eastern","mic":"OTCM"}, ...
>>> r.session.get('https://api.robinhood.com/markets/XNYS/hours/2015-09-10')
{"closes_at":"2015-09-10T20:00:00Z","next_open_hours":"https://api.robinhood.com/markets/XNYS/hours/2015-09-11/","previous_open_hours":"https://api.robinhood.com/markets/XNYS/hours/2015-09-09/","is_open":true,"date":"2015-09-10","opens_at":"2015-09-10T13:30:00Z"}
```
- Downloading of statements as PDFs:
```python
>>> r.session.get(r.endpoints['edocuments']).json()
[{"account":"https://api.robinhood.com/accounts/2PY73824/","url":"https://api.robinhood.com/documents/22d8ac82-38ca-7483-93ef-42e12e8cae93/","created_at":"2015-09-02T02:21:42.468680Z","updated_at":"2015-09-08T07:18:21.969094Z","download_url":"https://api.robinhood.com/documents/22d8ac82-38ca-7483-93ef-42e12e8cae93/download/","insert_3_url":"","insert_4_url":"","insert_1_url":"","insert_6_url":"","date":"2015-08-31","insert_2_url":"","type":"account_statement","insert_5_url":""}, ...
# Visiting res[0].account["download_url"] redirects to an s3 hosted PDF,
# even without credentials, so this method could write to disk, or just
# open a chrome window for viewing / download.
```
- A way to run place_buy_order() (or a dummy method) and receive a proper response without actually placing an order so that developers can build applications without having to place tons of live orders during development and testing.

If any of those sound interesting, feel free to jump in. You're also welcome to write up a suite of unit tests. Keep me in the loop regarding what you're planning to work on and how you're planning to implement to avoid duplicating work and to ease merging.

**Comment by benkroop** [comment]
Oh, and one more feature that needs to be completed is the ability to place stop limit and stop loss orders. The code is essentially written, but I couldn't figure out the order_type keyword to place them.

Take a look at lines 100 - 114 of robinhood.py:

```python
if order_type == "market":
    data += "&price=%f" % (float(self.get_quote(instrument['symbol'])[0]['bid_price']))
if order_type == "limit":
    data += "&price=%f" % (float(bid_price))
##Stop Loss and Stop Limit orders are a work in progress
##if order_type == "stop_loss":
##  data += "&stop_price=%f" % (float(stop_price))
##if order_type == "stop_limit":
##  data += "&price=%f&stop_price=%f" % (float(bid_price), float(stop_price))
```

order_types 'market' and 'limit' both work, and I'm 98% sure that "stop_price" is a valid and correct keyword, but I just haven't been able to figure out the stop keywords yet. If anyone has any guesses or their own strategies for figuring it out, give it a try and either let me know if you get it, or submit a pull request.

**Comment by frankV** [comment]
:+1: for this
I see this becoming a popular project. So it would be nice even for the sole purpose of having a build badge and coverage percentage. It's reassuring to some.

**Comment by benkroop** [comment]
After class today I'll have a discussion about options for CI. I'm with you on the advantages of having a build badge and coverage percentage.
